File: tests/js/test_tphcm_land_price_calculator.js
'use strict';

const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');
const vm = require('node:vm');

const root = path.resolve(__dirname, '..', '..');
const source = fs.readFileSync(
  path.join(root, 'static', 'js', 'tphcm_land_price_calculator.js'),
  'utf8'
);
const window = {};
vm.runInNewContext(source, { window, Intl, Number, String, Math });

const calculator = window.RadarLandPriceCalculator;
assert.ok(calculator, 'calculator must be exposed on window');

assert.deepEqual(
  JSON.parse(JSON.stringify(calculator.buildPayload({
    rowKey: 'row-1',
    landArea: '100',
    frontage: '5',
    depth: '20',
    mode: 'standard',
    access: 'alley',
    alleyWidth: '4',
    alleySurface: 'dirt',
    roadDistance: '100',
  }))),
  {
    row_key: 'row-1',
    land_area_m2: '100',
    frontage_m: '5',
    depth_m: '20',
    location: {
      mode: 'standard',
      access: 'alley',
      alley_min_width_m: '4',
      alley_surface: 'dirt',
      distance_to_named_road_m: '100',
    },
  }
);

assert.deepEqual(
  JSON.parse(JSON.stringify(calculator.buildPayload({
    rowKey: 'row-2',
    landArea: '80',
    frontage: '4',
    depth: '20',
    mode: 'multiple_frontages',
    access: 'alley',
    alleyWidth: '3',
    alleySurface: 'dirt',
    roadDistance: '120',
  }))),
  {
    row_key: 'row-2',
    land_area_m2: '80',
    frontage_m: '4',
    depth_m: '20',
    location: {
      mode: 'multiple_frontages',
    },
  }
);

assert.deepEqual(
  JSON.parse(JSON.stringify(calculator.buildPayload({
    rowKey: 'row-mixed',
    landArea: '500',
    frontage: '10',
    depth: '50',
    mode: 'standard',
    access: 'frontage',
    mixedMode: true,
    residentialArea: '100',
    agriculturalArea: '400',
    agriculturalType: 'perennial',
    agriculturalPosition: '1',
    inResidentialArea: false,
    sameParcelHasHouse: true,
    residentialUseCustom: true,
    residentialFrontage: '5',
    residentialDepth: '20',
  }))),
  {
    row_key: 'row-mixed',
    parcel_mode: 'mixed',
    land_area_m2: '500',
    frontage_m: '10',
    depth_m: '50',
    residential_area_m2: '100',
    agricultural_area_m2: '400',
    residential_geometry: {
      use_custom: true,
      frontage_m: '5',
      depth_m: '20',
    },
    location: {
      mode: 'standard',
      access: 'frontage',
    },
    agricultural: {
      land_type: 'perennial',
      position: '1',
      in_residential_area: false,
      same_parcel_has_house: true,
    },
  }
);

const html = calculator.renderResult({
  position: {
    label: 'Vị trí 3 <script>alert(1)</script>',
    factor: 0.288,
    breakdown: [{ label: 'Hẻm đất', factor: 0.8 }],
  },
  geometry: { legal_area_m2: 100, mismatch_warning: true },
  values: {
    residential: {
      base_unit_price: 10_000_000,
      average_unit_price: 2_880_000,
      total_value: 288_000_000,
      bands: [{
        code: 'front',
        area_m2: 100,
        factor: 1,
        unit_price: 2_880_000,
        subtotal: 288_000_000,
      }],
    },
  },
  warnings: [{
    code: 'geometry_mismatch',
    message: 'Cần đối chiếu hình thể thửa.',
  }],
});

assert.match(html, /2,88 triệu\/m²/);
assert.match(html, /288 triệu/);
assert.match(html, /Cần đối chiếu hình thể thửa/);
assert.match(html, /Hẻm đất/);
assert.match(html, /Phần phía trước/);
assert.doesNotMatch(html, /<script>/);
assert.match(html, /&lt;script&gt;/);

const emptyHtml = calculator.renderResult({
  position: { label: 'Vị trí 1', factor: 1, breakdown: [] },
  geometry: { legal_area_m2: 100 },
  values: {
    residential: {
      base_unit_price: null,
      average_unit_price: null,
      total_value: null,
      bands: [],
    },
  },
  warnings: [],
});
assert.match(emptyHtml, /Không áp dụng/);

const mixedHtml = calculator.renderResult({
  parcel_mode: 'mixed',
  position: { label: 'Vị trí 1', factor: 1, breakdown: [] },
  geometry: { legal_area_m2: 500 },
  mixed_use: {
    total_area_m2: 500,
    residential: {
      area_m2: 100,
      assumption: 'front_strip',
      base_unit_price: 687_200_000,
      average_unit_price: 687_200_000,
      total_value: 68_720_000_000,
      bands: [{
        code: 'front',
        area_m2: 100,
        factor: 1,
        unit_price: 687_200_000,
        subtotal: 68_720_000_000,
      }],
    },
    agricultural: {
      area_m2: 400,
      land_type: 'perennial',
      land_type_label: 'Đất trồng cây lâu năm',
      zone: 1,
      position: 1,
      pricing_mode: 'article_5_8',
      normal_unit_price: 1_440_000,
      special_unit_price: 68_720_000,
      unit_price: 68_720_000,
      total_value: 27_488_000_000,
      floor_applied: false,
      cap_applied: false,
      manual_review_required: false,
      formula: [
        'Áp dụng khoản 8 Điều 5.',
        '<script>alert(1)</script>',
      ],
    },
    total_value: 96_208_000_000,
  },
  warnings: [{
    code: 'residential_front_strip_assumption',
    message: 'Giả định phần đất ở nằm phía trước.',
  }],
});

assert.match(mixedHtml, /Tổng giá trị theo bảng Nhà nước/);
assert.match(mixedHtml, /96,208 tỷ/);
assert.match(mixedHtml, /Đất trồng cây lâu năm/);
assert.match(mixedHtml, /Vùng I/);
assert.match(mixedHtml, /Vị trí 1/);
assert.match(mixedHtml, /Khoản 8 Điều 5/);
assert.match(mixedHtml, /68,72 triệu\/m²/);
assert.match(mixedHtml, /Giả định phần đất ở nằm phía trước/);
assert.doesNotMatch(mixedHtml, /<script>/);
assert.match(mixedHtml, /&lt;script&gt;/);

const manualMixedHtml = calculator.renderResult({
  parcel_mode: 'mixed',
  position: { label: 'Vị trí 1', factor: 1, breakdown: [] },
  geometry: { legal_area_m2: 100 },
  mixed_use: {
    total_area_m2: 100,
    residential: {
      area_m2: 50,
      base_unit_price: 10_000_000,
      average_unit_price: 10_000_000,
      total_value: 500_000_000,
      bands: [],
    },
    agricultural: {
      area_m2: 50,
      land_type_label: 'Đất nông nghiệp khác',
      zone: 3,
      position: 1,
      pricing_mode: 'manual_review',
      unit_price: null,
      total_value: null,
      manual_review_required: true,
      formula: ['Cần đối chiếu loại đất liền kề.'],
    },
    total_value: null,
  },
  warnings: [],
});

assert.match(manualMixedHtml, /Chưa thể tính tổng/);
assert.match(manualMixedHtml, /Cần đối chiếu loại đất liền kề/);

File: tests/test_tphcm_land_price_calculator.py
from decimal import Decimal

import pytest

from services.tphcm_land_price_calculator import (
    build_depth_bands,
    calculate_land_price,
    calculate_mixed_land_price,
    resolve_location,
)


@pytest.mark.parametrize(
    ("width", "position", "factor"),
    [
        ("2.99", 4, Decimal("0.32")),
        ("3", 3, Decimal("0.40")),
        ("4.99", 3, Decimal("0.40")),
        ("5", 2, Decimal("0.50")),
    ],
)
def test_alley_width_boundaries_select_official_position(width, position, factor):
    result = resolve_location(
        {
            "mode": "standard",
            "access": "alley",
            "alley_min_width_m": width,
            "alley_surface": "paved",
            "distance_to_named_road_m": "20",
        }
    )

    assert result["position"] == position
    assert result["factor"] == factor


def test_dirt_and_one_hundred_meter_factors_are_explainable():
    result = resolve_location(
        {
            "mode": "standard",
            "access": "alley",
            "alley_min_width_m": "4",
            "alley_surface": "dirt",
            "distance_to_named_road_m": "100",
        }
    )

    assert result["position"] == 3
    assert result["factor"] == Decimal("0.288")
    assert result["breakdown"] == [
        {"code": "position_3", "label": "Vị trí 3", "factor": Decimal("0.40")},
        {"code": "dirt_alley", "label": "Hẻm đất", "factor": Decimal("0.80")},
        {
            "code": "distance_100m",
            "label": "Cách đường có tên từ 100m",
            "factor": Decimal("0.90"),
        },
    ]


def test_distance_below_one_hundred_meters_is_not_discounted():
    result = resolve_location(
        {
            "mode": "standard",
            "access": "alley",
            "alley_min_width_m": "4",
            "alley_surface": "paved",
            "distance_to_named_road_m": "99.99",
        }
    )

    assert result["factor"] == Decimal("0.40")


@pytest.mark.parametrize(
    ("mode", "factor"),
    [
        ("multiple_frontages", Decimal("1.10")),
        ("special_seventy_percent", Decimal("0.70")),
    ],
)
def test_special_mode_replaces_standard_alley_factors(mode, factor):
    result = resolve_location(
        {
            "mode": mode,
            "access": "alley",
            "alley_min_width_m": "2",
            "alley_surface": "dirt",
            "distance_to_named_road_m": "200",
        }
    )

    assert result["factor"] == factor
    assert len(result["breakdown"]) == 1


def test_residential_depth_bands_use_five_and_eight_frontage_multiples():
    bands = build_depth_bands("500", "5", "100", "residential")

    assert bands == [
        {"code": "front", "area_m2": Decimal("125"), "factor": Decimal("1.00")},
        {"code": "middle", "area_m2": Decimal("75"), "factor": Decimal("0.80")},
        {"code": "rear", "area_m2": Decimal("300"), "factor": Decimal("0.70")},
    ]


def test_commerce_depth_bands_use_two_and_four_frontage_multiples():
    bands = build_depth_bands("500", "5", "100", "commerce_service")

    assert bands == [
        {"code": "front", "area_m2": Decimal("50"), "factor": Decimal("1.00")},
        {"code": "middle", "area_m2": Decimal("50"), "factor": Decimal("0.60")},
        {"code": "rear", "area_m2": Decimal("400"), "factor": Decimal("0.40")},
    ]


def test_calculation_returns_average_and_total_for_each_land_type():
    result = calculate_land_price(
        {
            "residential": 10_000,
            "commerce_service": 6_000,
            "production_business": 4_000,
        },
        land_area_m2="100",
        frontage_m="5",
        depth_m="20",
        location={"mode": "standard", "access": "frontage"},
    )

    assert result["values"]["residential"]["average_unit_price"] == 10_000_000
    assert result["values"]["residential"]["total_value"] == 1_000_000_000
    assert (
        result["values"]["commerce_service"]["average_unit_price"] == 4_800_000
    )
    assert result["values"]["commerce_service"]["total_value"] == 480_000_000
    assert (
        result["values"]["production_business"]["average_unit_price"] == 3_200_000
    )
    assert result["values"]["production_business"]["total_value"] == 320_000_000


def test_geometry_mismatch_over_ten_percent_adds_warning():
    result = calculate_land_price(
        {
            "residential": 10_000,
            "commerce_service": 6_000,
            "production_business": 4_000,
        },
        land_area_m2="130",
        frontage_m="5",
        depth_m="20",
        location={"mode": "standard", "access": "frontage"},
    )

    assert result["geometry"]["mismatch_warning"] is True
    assert result["warnings"][0]["code"] == "geometry_mismatch"
    assert sum(
        band["area_m2"] for band in result["values"]["residential"]["bands"]
    ) == Decimal("130")


def test_mixed_calculation_adds_residential_and_agricultural_values():
    result = calculate_mixed_land_price(
        {
            "residential": 10_000,
            "commerce_service": 6_000,
            "production_business": 4_000,
        },
        area_name="XÃ CỦ CHI",
        land_area_m2="500",
        frontage_m="10",
        depth_m="50",
        residential_area_m2="100",
        agricultural_area_m2="400",
        residential_geometry={"use_custom": False},
        location={"mode": "standard", "access": "frontage"},
        agricultural={
            "land_type": "perennial",
            "position": 1,
            "in_residential_area": False,
            "same_parcel_has_house": False,
        },
    )

    assert result["parcel_mode"] == "mixed"
    assert result["mixed_use"]["residential"]["assumption"] == "front_strip"
    assert result["mixed_use"]["residential"]["geometry"]["frontage_m"] == Decimal(
        "10"
    )
    assert result["mixed_use"]["residential"]["geometry"]["depth_m"] == Decimal(
        "10"
    )
    assert result["mixed_use"]["residential"]["total_value"] == 1_000_000_000
    assert result["mixed_use"]["agricultural"]["unit_price"] == 840_000
    assert result["mixed_use"]["agricultural"]["total_value"] == 336_000_000
    assert result["mixed_use"]["total_value"] == 1_336_000_000


def test_mixed_calculation_accepts_split_within_one_hundredth_square_meter():
    result = calculate_mixed_land_price(
        {"residential": 10_000},
        area_name="XÃ CỦ CHI",
        land_area_m2="100",
        frontage_m="5",
        depth_m="20",
        residential_area_m2="40",
        agricultural_area_m2="59.99",
        residential_geometry={"use_custom": False},
        location={"mode": "standard", "access": "frontage"},
        agricultural={"land_type": "annual", "position": 1},
    )

    assert result["mixed_use"]["split_difference_m2"] == Decimal("0.01")


def test_mixed_calculation_rejects_area_split_outside_tolerance():
    from services.tphcm_land_price_calculator import CalculationValidationError

    with pytest.raises(CalculationValidationError) as exc_info:
        calculate_mixed_land_price(
            {"residential": 10_000},
            area_name="XÃ CỦ CHI",
            land_area_m2="100",
            frontage_m="5",
            depth_m="20",
            residential_area_m2="40",
            agricultural_area_m2="59.98",
            residential_geometry={"use_custom": False},
            location={"mode": "standard", "access": "frontage"},
            agricultural={"land_type": "annual", "position": 1},
        )

    assert "agricultural_area_m2" in exc_info.value.field_errors


def test_custom_residential_geometry_keeps_legal_area_and_warns_on_mismatch():
    result = calculate_mixed_land_price(
        {"residential": 10_000},
        area_name="XÃ CỦ CHI",
        land_area_m2="500",
        frontage_m="10",
        depth_m="50",
        residential_area_m2="100",
        agricultural_area_m2="400",
        residential_geometry={
            "use_custom": True,
            "frontage_m": "5",
            "depth_m": "30",
        },
        location={"mode": "standard", "access": "frontage"},
        agricultural={"land_type": "annual", "position": 1},
    )

    residential = result["mixed_use"]["residential"]
    assert residential["assumption"] == "custom_geometry"
    assert residential["geometry"]["legal_area_m2"] == Decimal("100")
    assert residential["geometry"]["mismatch_warning"] is True
    assert any(
        warning["code"] == "residential_geometry_mismatch"
        for warning in result["warnings"]
    )


def test_other_agricultural_prevents_a_misleading_combined_total():
    result = calculate_mixed_land_price(
        {"residential": 10_000},
        area_name="XÃ CỦ CHI",
        land_area_m2="500",
        frontage_m="10",
        depth_m="50",
        residential_area_m2="100",
        agricultural_area_m2="400",
        residential_geometry={"use_custom": False},
        location={"mode": "standard", "access": "frontage"},
        agricultural={"land_type": "other_agricultural", "position": 1},
    )

    assert result["mixed_use"]["agricultural"]["manual_review_required"] is True
    assert result["mixed_use"]["total_value"] is None
